zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java

```diff
@@ -1256,21 +1271,82 @@ private void reconnect() {
     }
   }
 
+
+
+  private void doAsyncSync(final ZooKeeper zk, final String path, final long startT,
+      final ZkAsyncCallbacks.SyncCallbackHandler cb) {
+    zk.sync(path, cb,
+        new ZkAsyncRetryCallContext(_asyncCallRetryThread, cb, _monitor, startT, 0, true) {
+          @Override
+          protected void doRetry() throws Exception {
+            doAsyncSync(zk, path, System.currentTimeMillis(), cb);
+          }
+        });
+  }
+
+  /*
+   *  Note, issueSync takes a ZooKeeper (client) object and pass it to doAsyncSync().
+   *  The reason we do this is that we want to ensure each new session event is preceded with exactly
+   *  one sync() to server. The sync() is to make sure the server would not see stale data.
+   *
+   *  ZooKeeper client object has an invariant of each object has one session. With this invariant
+   *  we can achieve each one sync() to server upon new session establishment. The reasoning is:
+   *  issueSync() is called when fireNewSessionEvents() which in under eventLock of ZkClient. Thus
+   *  we are guaranteed the ZooKeeper object passed in would have the new incoming sessionId. If by
+   *  the time sync() is invoked, the session expires. The sync() would fail with a stale session.
+   *  This is exactly what we want. The newer session would ensure another fireNewSessionEvents.
+   */
+  private boolean issueSync(ZooKeeper zk) {
```

Comment:
If we do not plan to change doAsynSync as a public API, i.e, it will use internally in this case, why not merge this two methods together to one?

Reply:
doAsyncSync() follows the pattern of other doAsync{verb} in terms of threading pattern. 

IssueSync actually waits for the result of doAsyncSync(). See the waitForSuccess().

Seems to me It is not easy to merge them together, or maybe I did not fully grab what you mean here. So maybe you can scratch the code of how to do it?